modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java

```diff
@@ -0,0 +1,273 @@
+                        .whenComplete((unused, throwable) -> {
+                            if (throwable != null) {
+                                LOG.error("Index build error: [{}]", throwable, createCommonTableIndexInfo());
+                            }
+                        });
+            } catch (Throwable t) {
+                LOG.error("Index build error: [{}]", t, createCommonTableIndexInfo());
+            } finally {
+                busyLock.leaveBusy();
+            }
+        }
+
+        private boolean isLocalNodeLeader(RaftGroupService raftGroupService) {
+            Peer leader = raftGroupService.leader();
+
+            assert leader != null : "tableId=" + table.tableId() + ", partitionId=" + partitionId;
+
+            return localNodeConsistentId().equals(leader.consistentId());
+        }
+
+        private List<RowId> createBatchRowIds(RowId lastBuiltRowId, int batchSize) {
+            MvPartitionStorage mvPartition = table.internalTable().storage().getMvPartition(partitionId);
+
+            assert mvPartition != null : createCommonTableIndexInfo();
+
+            List<RowId> batch = new ArrayList<>(batchSize);
+
+            for (int i = 0; i < batchSize && lastBuiltRowId != null; i++) {
+                lastBuiltRowId = mvPartition.closestRowId(lastBuiltRowId);
+
+                if (lastBuiltRowId == null) {
+                    break;
+                }
+
+                batch.add(lastBuiltRowId);
+
+                lastBuiltRowId = lastBuiltRowId.increment();
+            }
+
+            return batch;
+        }
+
+        private BuildIndexCommand createBuildIndexCommand(List<RowId> rowIds, boolean finish) {
+            return TABLE_MESSAGES_FACTORY.buildIndexCommand()
+                    .tablePartitionId(TABLE_MESSAGES_FACTORY.tablePartitionIdMessage()
+                            .tableId(table.tableId())
+                            .partitionId(partitionId)
+                            .build()
+                    )
+                    .indexId(tableIndexView.id())
+                    .rowIds(rowIds.stream().map(RowId::uuid).collect(toList()))
+                    .finish(finish)
+                    .build();
+        }
+
+        private String createCommonTableIndexInfo() {
+            return "table=" + table.name() + ", tableId=" + table.tableId()
+                    + ", partitionId=" + partitionId
+                    + ", index=" + tableIndexView.name() + ", indexId=" + tableIndexView.id();
+        }
+
+        private String localNodeConsistentId() {
+            return clusterService.topologyService().localMember().name();
+        }
+
+        private @Nullable RowId getNextRowIdForNextBatch(List<RowId> batch) {
+            return batch.isEmpty() ? null : batch.get(batch.size() - 1).increment();
+        }
+
+        private @Nullable List<RowId> collectRowIdBatch() {
+            RowId nextRowIdToBuilt;
+
+            if (nextRowIdToBuiltFromPreviousBatch == null) {
```

Comment:
As I already mentioned, I don't want to see this null-check, I'd prefer this value to be non-null

Reply:
I wrote above why I don’t like this approach and it doesn’t fit.

Reply:
I don't agree with the statement that it doesn't fit. I think you start tasks at inappropriate time, let's discuss it further